src/Microsoft.DotNet.TestFramework/TestAssetInfo.cs

```diff
@@ -72,6 +96,35 @@ public TestAssetInstance CreateInstance([CallerMemberName] string callingMethod
             return testInstance;
         }
 
+        internal IEnumerable<FileInfo> GetSourceFiles()
+        {
+            ThrowIfAssetSourcesHaveChanged();
+            
+            return GetInventory(_inventoryFiles.Source, null, () => {});
+        }
+
+        internal IEnumerable<FileInfo> GetRestoreFiles()
+        {
+            ThrowIfAssetSourcesHaveChanged();
+            
+            return GetInventory(_inventoryFiles.Restore, GetSourceFiles, DoRestoreWithLock);
+        }
+
+        internal IEnumerable<FileInfo> GetBuildFiles()
+        {
+            ThrowIfAssetSourcesHaveChanged();
+            
+            return GetInventory(
+                _inventoryFiles.Build,
+                () =>
+                {
+                    var preInventory = new List<FileInfo>(GetRestoreFiles());
```

Comment:
nit: Why use the constructor for one and `AddRange` for the other? The slight inconsistency is a tiny piece of readability friction.

Entirely a matter of personal taste but:

```
return GetInventory(
                _inventoryFiles.Build,
                () => GetRestoreFiles()
                          .Concat(GetSourceFiles())
                          .ToList(),
                DoBuildWithLock);
```